### ISSUE TITLE
Record example doesn't seem to be working

### ISSUE THREAD

**Issue by djfm** [RASPIAUDIO/Muse_library#1]
Hi,

First of all it's nice to see a proper SDK for the Muse taking shape, thanks!

I got the audio output working properly thanks to the [Audio_out example](https://github.com/RASPIAUDIO/Muse_library/blob/main/examples/LUXE/Audio_out/Audio_out.ino), I'm now trying to record audio taking inspiration from the [record1 example](https://github.com/RASPIAUDIO/Muse_library/blob/main/examples/LUXE/record1/record1.ino), but unfortunately it seems to me there is no `I2SClass` defined anywhere, and the `ESP_I2S.h` header isn't found either. I would have expected to find it in the https://github.com/schreibfaul1/ESP32-audioI2S dependency, but as far as I can tell it isn't defined there either.

It seems the whole thing is under active development so I wouldn't be terribly surprised if all of the examples didn't work perfectly yet, but am I missing something? What's the currently recommended way to record audio using the Muse Luxe?

I have the three dependencies of `Muse_library` listed in the [README](https://github.com/RASPIAUDIO/Muse_library) installed, and I've installed everything directly from GitHub using the main branch of each repository if that matters.

Thanks in advance for any help you might be able to provide.

**Comment by RASPIAUDIO** [comment]
Hi 
Have you try to install some of the last esp32 SDK in 3.x ?
Go in the Arduino board manager

**Comment by djfm** [comment]
Hello, thanks for your reply,

By esp32 SDK do you mean [arduino-esp32](https://github.com/espressif/arduino-esp32)?

If so, then yes, I have version `3.20017.0`.

There is a class named `I2SClass` in `<I2S.h>`, but trying to include it together with `"museluxe.h"` results in a ton of compiler errors along the lines of:

```
framework-arduinoespressif32/libraries/I2S/src/I2S.h:61:3: error: conflicting declaration 'typedef enum i2s_mode_t i2s_mode_t'
 } i2s_mode_t;
```

and the class's API is also vastly different from that of the example.

I'm using PlatformIO btw, but I don't think it changes anything here.

Which board are you using in the IDE?

EDIT: Ah, I'm realizing I might actually have a version of `arduino-esp32` which is too recent, I'll keep you posted.

**Comment by djfm** [comment]
Well, İ'm not quite sure what PlatformIO is doing right now actually, but it's not using the espressif libs I thought it was, so I'm closing this issue as it seems clear at this point that the problem is on my side.

If by chance you have an example `platformio.ini` configuration file that works for the muse it would be nice though.

**Comment by djfm** [comment]
Alright, turns out PlatformIO (stable) doesn't support the latest espressif framework, but in case anyone is interested, this `platformio.ini` works fine to get version 3.0.5 to work:

```
[env:esp32dev]
; For the platform below, as depending on a develop branch is risky,
; the commit at the time of writing is: 0bdfa3070724e777cf5fb05ce6ab23bafcf085e5
; and things appear to be working fine.
platform = https://github.com/platformio/platform-espressif32.git#develop
platform_packages =
    platformio/framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#3.0.7
	platformio/framework-arduinoespressif32-libs @ https://github.com/espressif/esp32-arduino-libs.git#idf-release/v5.1
board = esp32dev
framework = arduino
monitor_speed = 115200
board_build.partitions = huge_app.csv
lib_extra_dirs = lib
```

(`lib` would be the place where you put your dependencies such as `Muse_library` etc.)

That being said, the record example still doesn't work for me even though everything compiles fine and I'm able to stream mp3's using `Muse_library` without issues.

I'm not reopening the issue, I'll do some more digging, but would you mind confirming if the record example is working for you please?

**Comment by RASPIAUDIO** [comment]
Try wrover board and resync to the last version of our lib as we are currently working on it

**Comment by djfm** [comment]
Thank you, this is encouraging! I've updated the dependencies and changed the i2s init to mono like you did in the examples. In my case I'm trying to use both Audio to play remote files and i2s to record and send the sound to a server. With some fiddling I can get something which starts to sound like something recognizable on the server although it's still heavily saturated. The impact of `microphone_level` seems kinda random to me at the moment, I'll keep trying and follow the library's development. I did not notice a change when using wrover instead of esp32dev though.

**Comment by RASPIAUDIO** [comment]
Are you using both recording and playback simultaneously? This can be tricky, but we have managed it before in our factory tests. Check the code for the Muse Luxe in the Simple Bluetooth repository

**Comment by djfm** [comment]
Thanks for the advice, sorry my reply wasn't very clear, no I'm not playing
and recording simultaneously: I'm alternating between playing and recording
but in the same firmware.

I got it to work rather fine, I found I get the best results by muting the
ES8388 driver and calling `audio.stopSong()` before I start i2s for
recording, and conversely when I switch from recording to playing I stop
the i2s interface and call `audio.setPins` again and unmute the ES8388. I'm
not totally sure that all these steps are necessary but it seemed safer to
me this way and it doesn't create any noticeable delays so far.

I initially thought the recorded sound was distorted but it turned out it
was the server to which I stream the recorded audio which was not handling
the data correctly and not the microphone which was not recording it
properly. So it's all good now.

I feel like the microphone volume doesn't increase much despite using high
values in `es.microphone_volume()`, but it's loud enough to be used. Worst
case scenario I found out that I can boost the volume in software quite a
lot without degrading the quality.
